stdlib/string.gr

```diff
@@ -519,3 +519,596 @@ export let endsWith = (p: String, s: String) => {
+        codePoint = byte & 0x7n
+      } else {
+        throw MalformedUnicode
+      }
+      continue
+    }
+    if (!(lowerBoundary <= byte && byte <= upperBoundary)) {
+      throw MalformedUnicode
+    }
+    lowerBoundary = 0x80n
+    upperBoundary = 0xBFn
+    codePoint = (codePoint << 6n) | (byte & 0x3Fn)
+    bytesSeen += 1n
+    if (bytesSeen == bytesNeeded) {
+      result = codePoint
+      break
+    }
+  }
+  result: WasmI32
+}
+
+let _BYTES_OFFSET = 8n;
+
+// Encodes the given string using the given encoding scheme
+// @param s: String - The input string
+// @param encoding: Encoding - The encoding to use
+// @param includeBom: Bool - Whether to include the byte-order marker in the encoded output
+// @returns Bytes
+@disableGC
+export let encode = (s: String, encoding: Encoding, includeBom: Bool) => {
+  let (>>>) = WasmI32.shrU
+  let (-) = WasmI32.sub
+  let (&) = WasmI32.and
+  let (<) = WasmI32.ltU
+  let (<=) = WasmI32.leU
+  let (==) = WasmI32.eq
+
+  let size = encodedLength(s, encoding) + if (includeBom) {
+    match(encoding) {
+      UTF8 => 3,
+      UTF16_LE => 2,
+      UTF16_BE => 2,
+      UTF32_LE => 4,
+      UTF32_BE => 4,
+    }
+  } else 0
+  let (+) = WasmI32.add
+  let byteSize = WasmI32.fromGrain(byteLength(s)) >>> 1n
+  let len = WasmI32.fromGrain(length(s)) >>> 1n
+
+  let s = WasmI32.fromGrain(s)
+
+  let mut ptr = s + 8n
+  let end = ptr + byteSize
+
+  // Allocate destination bytes
+  let bytes = allocateBytes(WasmI32.fromGrain(size) >>> 1n)
+  let mut bytesIdx = 0n
+
+  if (includeBom) {
+    match(encoding) {
+      UTF8 => {
+        WasmI32.store8(bytes + bytesIdx, 0xEFn, _BYTES_OFFSET)
+        WasmI32.store8(bytes + bytesIdx + 1n, 0xBBn, _BYTES_OFFSET)
+        WasmI32.store8(bytes + bytesIdx + 2n, 0xBFn, _BYTES_OFFSET)
+        bytesIdx += 3n
+      },
+      UTF16_BE => {
+        WasmI32.store8(bytes + bytesIdx, 0xFEn, _BYTES_OFFSET)
+        WasmI32.store8(bytes + bytesIdx + 1n, 0xFFn, _BYTES_OFFSET)
+        bytesIdx += 2n
+      },
+      UTF16_LE => {
+        WasmI32.store8(bytes + bytesIdx, 0xFFn, _BYTES_OFFSET)
+        WasmI32.store8(bytes + bytesIdx + 1n, 0xFEn, _BYTES_OFFSET)
+        bytesIdx += 2n
+      },
+      UTF32_BE => {
+        WasmI32.store8(bytes + bytesIdx, 0n, _BYTES_OFFSET)
+        WasmI32.store8(bytes + bytesIdx + 1n, 0n, _BYTES_OFFSET)
+        WasmI32.store8(bytes + bytesIdx + 2n, 0xFEn, _BYTES_OFFSET)
+        WasmI32.store8(bytes + bytesIdx + 3n, 0xFFn, _BYTES_OFFSET)
+        bytesIdx += 4n
+      },
+      UTF32_LE => {
+        WasmI32.store8(bytes + bytesIdx, 0xFFn, _BYTES_OFFSET)
+        WasmI32.store8(bytes + bytesIdx + 1n, 0xFEn, _BYTES_OFFSET)
+        WasmI32.store8(bytes + bytesIdx + 2n, 0n, _BYTES_OFFSET)
+        WasmI32.store8(bytes + bytesIdx + 3n, 0n, _BYTES_OFFSET)
+        bytesIdx += 4n
+      }
+    }
+  }
+
+  while (ptr < end) {
+    let byte = WasmI32.load8U(ptr, 0n)
+    // number of bytes in Grain UTF-8 string
+    // ptr+ offset + 8n
```

Comment:
I got pretty confused trying to understand what this comment was trying to say. I think this is more clear:
```suggestion
    // number of bytes spanning this UTF-8-encoded scalar value
```

Reply:
Sorry, this half-included some commented out copy-paste stuff.